File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/LineReader.java
package com.flat.simplex.parser.logic;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.parser.logic.line.*;
import com.flat.simplex.parser.logic.line.call.*;

import java.util.ArrayList;

public class LineReader {

    private Block parent;
    private Context context;

    private ArrayList<Line> calls = new ArrayList<>();
    private LineCall firstCall = null;
    private LineCall lastCall = null;

    private Token token;
    private Token end;

    public LineReader(Block parent) {
        this.parent = parent;
        this.context = parent.getContext();
    }

    public ArrayList<Line> read(Token tokenStart, Token tokenEnd) {
        this.token = tokenStart;
        this.end = tokenEnd;

        while (token != end && token != null) {
            Key key = token.getKey();
            if (key == Key.Word || key == Key.This || key == Key.Global) {
                consumeField();

            } else if (key == Key.Number || key == Key.String || key == Key.Undefined ||
                    key == Key.True || key == Key.False) {
                consumeValue();

            } else if (key == Key.Function) {
                consumeFunction();

            } else if (key == Key.Dot) {
                consumeDot();

            } else if (key == Key.Param) {
                consumeParam();

            } else if (key == Key.Index) {
                consumeIndex();

            } else if (key == Key.Brace) {
                consumeBrace();

            }  else if (key.op > 0) {
                consumeOp();

            } else {
                if (lastCall != null) {
                    addLastLineChain();
                }
                context.error(token, Error.unexpectedToken);
            }
            token = token.getNext();
        }
        if (lastCall != null) {
            addLastLineChain();
        }
        return calls;
    }

    private void consumeField() {
        if (lastCall != null) {
            addLastLineChain();
            context.error(token, Error.lineMissingAccessor);
        }
        bindNext(new CallField(parent, token));
    }

    private void consumeValue() {
        if (lastCall != null) {
            addLastLineChain();
            context.error(token, Error.lineMissingAccessor);
        }
        bindNext(new CallValue(parent, token));
    }

    private void consumeFunction() {
        Token fEnd = token.getNext();
        if (fEnd != null && fEnd != end && fEnd.getKey() == Key.Param) {
            token = fEnd;
            fEnd = fEnd.getNext();
            if (fEnd != null && fEnd != end && fEnd.getKey() == Key.Brace) {
                token = fEnd;
                fEnd = fEnd.getNext();
            }
        }
        if (lastCall != null) {
            addLastLineChain();
            context.error(token, Error.lineMissingAccessor);
        }
        bindNext(new CallFunction(parent, token, fEnd));
    }

    private void consumeDot() {
        if (lastCall == null) {
            context.error(token, Error.unexpectedToken);
        } else {
            if (token.getNext() != null && token.getNext() != end) {
                if (token.getNext().getKey() == Key.Word) {
                    token = token.getNext();
                    bindNext(new CallField(parent, token));
                } else {
                    addLastLineChain();
                    context.error(token, Error.unexpectedToken);
                }
            } else {
                addLastLineChain();
                context.error(token, Error.unexpectedEndOfTokens);
            }
        }
    }

    private void consumeParam() {
        if (lastCall == null) {
            bindNext(new CallGroup(parent, token));
        } else {
            bindNext(new CallMethod(parent, token));
        }
    }

    private void consumeIndex() {
        if (lastCall == null) {
            bindNext(new CallArray(parent, token));
        } else {
            bindNext(new CallIndexer(parent, token));
        }
    }

    private void consumeBrace() {
        if (lastCall != null) {
            addLastLineChain();
            context.error(token, Error.lineMissingAccessor);
        }
        bindNext(new CallStruct(parent, token));
    }

    private void consumeOp() {
        if (lastCall != null) {
            addLastLineChain();
        }
        addOp(new LineOp(parent, token));
    }

    private void bindNext(LineCall lineCall) {
        if (lastCall != null) {
            lastCall.setNext(lineCall);
        } else {
            firstCall = lineCall;
        }
        lastCall = lineCall;
    }

    private void addLastLineChain() {
        calls.add(new LineChain(firstCall));
        lastCall = null;
        firstCall = null;
    }

    private void addOp(LineOp lineOp) {
        calls.add(lineOp);
    }
}

File: SimplexApp/simplex/parser/logic/LineReader.js
var Key, LineOp, LineChain, CallArray, CallField, CallFunction, CallGroup, CallIndexer, CallMethod, CallStruct, CallValue, Error;

class LineReader {
    constructor(parent) {
        this.parent = parent;
        this.calls = [];
        this.firstCall = null;
        this.lastCall = null;
        this.token = null;
        this.end = null;

    }

    read(tokenStart, tokenEnd) {
        this.token = tokenStart;
        this.end = tokenEnd;
        while (this.token !== this.end && this.token !== null) {
            let key = this.token.getKey();
            if (key === Key.Word || key === Key.This || key === Key.Global) {
                this.consumeField();
            } else if (key === Key.Number || key === Key.String || key === Key.Undefined || key === Key.True || key === Key.False) {
                this.consumeValue();
            } else if (key === Key.Function) {
                this.consumeFunction();
            } else if (key === Key.Dot) {
                this.consumeDot();
            } else if (key === Key.Param) {
                this.consumeParam();
            } else if (key === Key.Index) {
                this.consumeIndex();
            } else if (key === Key.Brace) {
                this.consumeBrace();
            } else if (key.op > 0) {
                this.consumeOp();
            } else {
                if (this.lastCall !== null) {
                    this.addLastLineChain();
                }
                this.parent.error(this.token, Error.unexpectedToken);
            }
            this.token = this.token.getNext();
        }
        if (this.lastCall !== null) {
            this.addLastLineChain();
        }
        return this.calls;
    }

    consumeField() {
        if (this.lastCall !== null) {
            this.addLastLineChain();
            this.parent.error(this.token, Error.lineMissingAccessor);
        }
        this.bindNext(new CallField(this.parent, this.token));
    }

    consumeValue() {
        if (this.lastCall !== null) {
            this.addLastLineChain();
            this.parent.error(this.token, Error.lineMissingAccessor);
        }
        this.bindNext(new CallValue(this.parent, this.token));
    }

    consumeFunction() {
        let start = this.token;
        let fEnd = this.token.getNext();
        if (fEnd !== null && fEnd !== this.end && fEnd.getKey() === Key.Param) {
            this.token = fEnd;
            fEnd = fEnd.getNext();
            if (fEnd !== null && fEnd !== this.end && fEnd.getKey() === Key.Brace) {
                this.token = fEnd;
                fEnd = fEnd.getNext();
            }
        }
        if (this.lastCall !== null) {
            this.addLastLineChain();
            this.parent.error(this.token, Error.lineMissingAccessor);
        }
        this.bindNext(new CallFunction(this.parent, start, fEnd));
    }

    consumeDot() {
        if (this.lastCall === null) {
            this.parent.error(this.token, Error.unexpectedToken);

        } else if (this.token.getNext() !== null && this.token.getNext() !== this.end) {
            if (this.token.getNext().getKey() === Key.Word) {
                this.token = this.token.getNext();
                this.bindNext(new CallField(this.parent, this.token));
            } else {
                this.addLastLineChain();
                this.parent.error(this.token, Error.unexpectedToken);
            }

        } else {
            this.addLastLineChain();
            this.parent.error(this.token, Error.unexpectedEndOfTokens);

        }
    }

    consumeParam() {
        if (this.lastCall === null) {
            this.bindNext(new CallGroup(this.parent, this.token));
        } else {
            this.bindNext(new CallMethod(this.parent, this.token));
        }
    }

    consumeIndex() {
        if (this.lastCall === null) {
            this.bindNext(new CallArray(this.parent, this.token));
        } else {
            this.bindNext(new CallIndexer(this.parent, this.token));
        }
    }

    consumeBrace() {
        if (this.lastCall !== null) {
            this.addLastLineChain();
            this.parent.error(this.token, Error.lineMissingAccessor);
        }
        this.bindNext(new CallStruct(this.parent, this.token));
    }

    consumeOp() {
        if (this.lastCall !== null) {
            this.addLastLineChain();
        }
        this.addOp(new LineOp(this.parent, this.token));
    }

    bindNext(lineCall) {
        if (this.lastCall !== null) {
            this.lastCall.setNext(lineCall);
        } else {
            this.firstCall = lineCall;
        }
        this.lastCall = lineCall;
    }

    addLastLineChain() {
        this.calls.push(new LineChain(this.firstCall));
        this.lastCall = null;
        this.firstCall = null;
    }

    addOp(lineOp) {
        this.calls.push(lineOp);
    }
}

module.exports = LineReader;
Key = require("simplex/lexer/Key.js");
LineOp = require("simplex/parser/logic/line/LineOp.js");
LineChain = require("simplex/parser/logic/line/LineChain.js");
CallArray = require("simplex/parser/logic/line/call/CallArray.js");
CallField = require("simplex/parser/logic/line/call/CallField.js");
CallFunction = require("simplex/parser/logic/line/call/CallFunction.js");
CallGroup = require("simplex/parser/logic/line/call/CallGroup.js");
CallIndexer = require("simplex/parser/logic/line/call/CallIndexer.js");
CallMethod = require("simplex/parser/logic/line/call/CallMethod.js");
CallStruct = require("simplex/parser/logic/line/call/CallStruct.js");
CallValue = require("simplex/parser/logic/line/call/CallValue.js");
Error = require("simplex/parser/logic/error/Error.js");